### PR TITLE
Added precise decimal multiplication and division

### LINKED ISSUE #8049: Support Decimal multiplication saving result scale
В decimal есть довольно неудобный подвох - если хочешь произвести умножение 2 чисел, то невозможно это сделать не увеличивая precision (хочется умножение с округлением). 

В абсолютно типичном сценарии amount Decimal64(2), exchange_rate Decimal64(6) сейчас происходят переполнения, это неудобно, и такое расширение точности там совсем не нужно. МБ просто нужно добавить вариант умножения с округлением?

```sql
SELECT round(toDecimal64(1, 8) * toDecimal64(1000, 8), 8)

Received exception from server (version 19.18.1):
Code: 407. DB::Exception: Received from localhost:9000. DB::Exception: Decimal math overflow. 

SET decimal_check_overflow = 0;

SELECT round(toDecimal64(1, 8) * toDecimal64(1000, 8), 8);

┌─round(multiply(toDecimal64(1, 8), toDecimal64(1000, 8)), 8)─┐
│                                       -844.6744073700000000 │
└─────────────────────────────────────────────────────────────┘

1 rows in set. Elapsed: 0.047 sec.

SELECT toTypeName(toDecimal64(1, 3) * toDecimal64(1, 4)) AS x

Row 1:
──────
x: Decimal(18, 7)
```
мб ему в таких случаях стоит разрешить расширять тип самостоятельно - до Decimal128?

> МБ просто нужно добавить вариант умножения с округлением?

Да, и это будет самым логичным и ожидаемым поведением. Но не ясно, насколько сложно сделать. Хотя если мы умеем определять overflow, значит можно сделать в slow path, и тогда скорость работы обычных случаев не изменится.

> Как вариант - менять поведение обычного умножения на это умножение с округлением настройкой по умолнчанию?

Лучше без настроек.

> При делении он остаётся с типом левого аргумента

А здесь всё правильно, это самое логичное поведение.

> По-другому технически нерационально

По моему, можно сделать даже без компромиссов по производительности.

### LINKED ISSUE #40573: Add precise Decimal division
Right now, `Decimal` devision (#30341, #38054) can cause some pain. 
It works fine, but with strict limitations -- e.g. we cannot divide Decimals that have scale more than half of max scale (#39600).

Currently used division has good speed -- in fact, we just make 1 division (expensive) and multiplication (not so expensive), but precision is poor. 

I suggest adding separate function (smth like `preciseDecimalDiv(a: Decimal, b: Decimal, prcsn: UInt8)` with user-defined result precision) for decimal division (alongside "usual" one).
Maybe even we can replace currently existing division by this more precise version, but it will be like 10 times slower (more division operations are to be made), and it may be unacceptable if you don't need high precision.
That's why I suggest having 2 separate functions.

### PR DESCRIPTION
**Pull request by zvonand**
Closes #40573, closes #8049
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Added `multiplyDecimal` and `divideDecimal` functions for decimal operations with fixed precision

### CODE REVIEW

**Comment by alexey-milovidov** [comment]
Introducing a new set of functions is a heavy solution, it will introduce a burden for the users to memorize these new functions.
Do you consider any alternatives?

**Comment by zvonand** [comment]
Well, my thoughts are that currently existing binary arithmetic funcs are already heavily overloaded, and adding more complexity here would not work. It took me not a single hour to really get into what's happening there.
Moreover, (IMO) these functions shall stand alone from `traditional` operations, e.g. `/` -- they work in a different way (they need 3 operands) and they work really slow (the price we have to pay to have controlled precision and handle possible intermediate overflows). I'm speeding them up now, but difference will still be high -- I haven't finished yet, but I estimate them to work 15-20x slower.
I think that if user uses these, he needs to be explicitly aware of all the pecularities he will meet.

@alexey-milovidov which alternatives do you think will work? I don't really think that making `divide` and `multiply` act in two really different ways is a good idea.
We may make a link in docs on Decimal datatype: `if you want precise computations and you have no time limits, then you shall use `DivideDecimal` and `multiplyDecimal`. Otherwise it is recommended to use `/` and `*` operations respectively`

File: tests/queries/0_stateless/02475_precise_decimal_arithmetics.sql [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+SELECT divideDecimal(toDecimal32(123.123, 3), toDecimal128(11.4, 1), 2);
+SELECT divideDecimal(toDecimal128(-123.76, 2), toDecimal128(11.123456, 6));
+SELECT divideDecimal(toDecimal32(123.123, 3), toDecimal128(-11.4, 1), 2);
+SELECT divideDecimal(toDecimal32(-123.123, 3), toDecimal128(-11.4, 1), 2);
+
+SELECT multiplyDecimal(toDecimal64(123.76, 2), toDecimal128(11.123456, 6));
+SELECT multiplyDecimal(toDecimal32(123.123, 3), toDecimal128(11.4, 1), 2);
+SELECT multiplyDecimal(toDecimal64(-123.76, 2), toDecimal128(11.123456, 6));
+SELECT multiplyDecimal(toDecimal32(123.123, 3), toDecimal128(-11.4, 1), 2);
+SELECT multiplyDecimal(toDecimal32(-123.123, 3), toDecimal128(-11.4, 1), 2);
```
Comment (vdimir) [inline tests/queries/0_stateless/02475_precise_decimal_arithmetics.sql:30]: Consider adding a bit more tests with Nullable and non-const columns, like

```sql
SELECT multiplyDecimal(toNullable(toDecimal64(10, 1)), toDecimal64(1000, 5));
SELECT sum(multiplyDecimal(toDecimal64(number, 1)), toDecimal64(number, 5))) FROM numbers(1000);
```

File: src/Functions/FunctionsDecimalArithmetics.h [hunk 2]
```diff
@@ -0,0 +1,445 @@
...
+    extern const int ILLEGAL_COLUMN;
+    extern const int ILLEGAL_TYPE_OF_ARGUMENT;
+    extern const int NUMBER_OF_ARGUMENTS_DOESNT_MATCH;
+    extern const int ILLEGAL_DIVISION;
+}
+
+
+struct DecimalOpHelpers
+{
+    static std::vector<UInt8> multiply(const std::vector<UInt8> & num1, const std::vector<UInt8> & num2)
```
Comment (vdimir) [inline src/Functions/FunctionsDecimalArithmetics.h:32]: Let's add some comments about how these functions work and what is the expected number representation.

Also, it's unclear why we use Int32 and UInt16 for sizes and how it's guaranteed to fit (probably add some safety checks?).

**Review by vdimir — COMMENTED** [review]
Thank you for your contribution!

In general, the code looks good. See some small comments.

Since all tests passed, we can merge PR and make follow-up PRs to close these minor things, or we can do it in this PR. Let me know what is more convenient.

**Comment by zvonand** [comment]
Will fix in no time :)

File: src/Functions/FunctionsDecimalArithmetics.h [hunk 3]
```diff
@@ -0,0 +1,445 @@
...
+    }
+};
+
+
+struct DivideDecimalsImpl
+{
+    static constexpr auto name = "divideDecimal";
+
+    template <typename FirstType, typename SecondType>
+    static inline NO_SANITIZE_UNDEFINED Decimal256
```
[the remark this answers is not in the document]
Comment (zvonand) [reply src/Functions/FunctionsDecimalArithmetics.h]: Left this by mistake. Fixing.

**Comment by zvonand** [comment]
@den-crane fixed

**Comment by zvonand** [comment]
@vdimir all your comments are taken into account. test fails seem unrelated to this PR

**Review by vdimir — APPROVED** [review]
Thank you for your changes!

File: src/Functions/FunctionsDecimalArithmetics.cpp [hunk 4]
```diff
@@ -0,0 +1,17 @@
+#include <Functions/FunctionsDecimalArithmetics.h>
+#include <Functions/FunctionFactory.h>
+
+namespace DB
+{
+REGISTER_FUNCTION(DivideDecimals)
+{
+    factory.registerFunction<FunctionsDecimalArithmetics<DivideDecimalsImpl>>(Documentation(
+        "Decimal division with given precision. Slower than simple `divide`, but has controlled precision and no sound overflows"));
```
Comment (alexey-milovidov) [inline src/Functions/FunctionsDecimalArithmetics.cpp:9]: The documentation is not enough.
Every function should be in its own .cpp file.

Comment (zvonand) [reply src/Functions/FunctionsDecimalArithmetics.cpp:9]: Ok, will do it like this next time.

**Pull request merged**